### PR TITLE
PR Adds Coldcard

### PR DESCRIPTION
**Pull request by nvk**

### CODE REVIEW

**Comment by harding** [comment]
See @crwatkins's detailed evaluation at https://github.com/bitcoin-dot-org/bitcoin.org/pull/2612#issuecomment-433630363

From that PR, it looks like the remaining issues are:

1. Confirmation that issues with the keypad have been resolved.  From Craig's evaluation: "The keypad is extremely difficult to use. On my first usage, it took me a long time to set my initial PIN because the keypad was generating many digits that I did not enter. This was addressed with a firmware update that made the "phantom" presses less frequent but did not eliminate them and added a setting for keyboard sensitivity. While using the "Normal" setting I still get some phantoms, and sometimes I have the keyboard stop responding (I don't know if that is related or not) but a power cycle usually makes it somewhat better. In "Normal" mode, the keypad is difficult to use and requires long presses which sometimes don't register and sometimes register more than once. I've had the spurious presses actually navigate me to the signing menu while I wasn't touching the device. The keypad really has to work better. Both of the reviews linked above described similar issues with the keypad."

2. The description text should be improved.  "For style and consistency reasons, there should be no words in all capitals nor any exclamation marks, and the spellings "open-source" and "MicroPython" should be used. I would further recommend, but up to you, for readability that open-source be only mentioned once, backup be only mentioned once (sentences combined), and that "no app required" be removed or reworded because clearly Electrum is required. The score for transparency is checkgoodtransparencydeterministic but I don't see any deterministic build procedures such as Gitian or Docker, so I suspect you want to change it to checkpasstransparencyopensource."

3. A rule forbidding sending transactions that pay more than 10% in fees should be made optional.  From the above thread, it seems like this will be released in a forthcoming firmware update.  If so, it's probably worth waiting until it is released.

My thanks to @crwatkins for taking the time to perform and writeup such a detailed review.

File: _translations/en.yml [hunk 1]
```diff
@@ -193,6 +193,7 @@ en:
     walletbitgo: "BitGo is a high-security multi-sig wallet, which protects your bitcoin from theft and loss. You maintain full custody; BitGo cannot spend or freeze funds. BitGo wallets are easy to use and offer advanced security features such as spending limits and multi-user access."
     walletgreenaddress: "GreenAddress is a user-friendly multi-signature wallet with improved security and privacy. At no time are your keys server side, even encrypted. For security reasons, you should always use 2FA and the browser extension or Android App."
     wallettrezor: "Trezor is a hardware wallet providing a high level of security without sacrificing convenience. Unlike cold storage, Trezor is able to sign transactions while connected to an online device. That means spending bitcoins is secure even when using a compromised computer."
+    walletcoldcard: "Coldcard is a easy to use, ultra-secure, opensource and affordable hardware wallet.  NO app required.  Easy back-up! MicroSD card slot for encrypted backups. Your private key is stored in a dedicated real crypto security chip. Open source software design runs Micropython and you can change it."
```
Comment (achow101) [inline _translations/en.yml]: As mentioned in other reviews, this should be changed to be consistent with the style of other descriptions.

I would recommend the following:

> The Coldcard is an easy to use, ultra-secure, open-source, and affordable hardware wallet with no app required to use. Encrypted backups are easy to do with the built in MicroSD card slot. Your private key is stored in a dedicated real crypto security chip. The open-source software design runs MicroPython and can be changed by you.

File: _wallets/coldcard.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    os:
+      - name: hardware
+        text: "walletcoldcard"
+        link: "https://coldcardwallet.com/"
+        source: "https://github.com/coldcard/"
+        screenshot: "coldcard.png"
+        check:
+          control: "checkgoodcontrolfull"
+          validation: "checkneutralvalidationvariable"
+          transparency: "checkgoodtransparencydeterministic"
```
Comment (achow101) [inline _wallets/coldcard.md]: I think this should be `checkpasstransparencyopensourcehardware`. I could not find any documentation that mentions that the firmware is deterministically built or any way to verify that the firmware on the device matches what you could build yourself.

**Review by achow101 — COMMENTED** [review]
> Confirmation that issues with the keypad have been resolved.

I've had a Coldcard myself and have been testing and using it as well. Since updating to firmware 1.0.2, I have not experienced spurious button presses. However button presses sometimes do not register. I don't think that this is necessarily a blocker for listing as it does not effect the security of the device.

>  that "no app required" be removed or reworded because clearly Electrum is required.

I think it's possible to just write a PSBT to a file, copy that file to a MicroSD card, then move the MicroSD card to the coldcard to sign it. So you wouldn't need Electrum, just any software that supports PSBT, e.g. Bitcoin Core.

> A rule forbidding sending transactions that pay more than 10% in fees should be made optional. From the above thread, it seems like this will be released in a forthcoming firmware update. If so, it's probably worth waiting until it is released.

I think it would be sufficient to mention this in the description. Perhaps we could add a new class of fee control check for where the fee is variable due to the software but also limited by hardware. Maybe `checkfailfeecontrolvariablelimited`? I think it would have to be a `checkfail` type though.

File: _translations/en.yml (hunk 1 above)
Comment (nvk) [reply _translations/en.yml]: 👍

File: _wallets/coldcard.md (hunk 2 above)
Comment (nvk) [reply _wallets/coldcard.md]: 👍

**Comment by nvk** [comment]
Changes made.

**Comment by nvk** [comment]
You can turn off Fee limiting the new firmware. (already available https://github.com/Coldcard/firmware/commit/aa263f0500be383a2f3f7faa432a9fcbe20ce2de)

**Comment by nvk** [comment]
Sorry, got the versions crossed. The new coming firmware upgrade has the option to turn OFF fee limiting. @peter-conalgo has not pushed them to the public repo yet.

**Comment by nvk** [comment]
Fee limiting code now public https://github.com/Coldcard/firmware/pull/11

**Comment by achow101** [comment]
utACK 5f200e1e128a2abbefe3f1c3cc7339c60b56f226

I think this is fine to merge, especially since the fee limiting will be switchable in a future release. Thoughts, @harding?

**Comment by harding** [comment]
I haven't used the wallet myself, so I don't feel qualified to ACK.  From the comments posted earlier, I think the "no app required" claim is misleading (the wallet isn't its own node or SPV client AFAICT, so you need to use some external app to send data to it and receive data from it) and I would personally prefer to see the firmware update to allow overriding the fee restriction be released and verified as working by a third party before listing the wallet.[1]

That said, @achow101, you've used the wallet and so you have a much better perspective about whether it's something we should be recommending as-is to the inexperienced users (or at least lesser-experienced users) of the Choose Your Wallet page.

[1] I'm going to be a harsh here, sorry, but I really don't like disallowing spends based on fee as a *percentage* of the total spend amount.  I'm totally ok with configurable restrictions on max fee similar to Bitcoin Core's `-maxtxfee`, but setting this to a percentage of the output amount both seems generally weird for Bitcoin and specifically makes it hard to test the wallet with small amounts of bitcoin---which is something any rational user is probably going to do before they deposit larger amounts.

**Comment by harding** [comment]
Oh, I just re-read the original review and @crwatkins is using the same firmware version as @achow101, so I think that's one user with the spurious keypresses issue unresolved and one with it resolved.  I did a quick Google search and I couldn't find anyone else specifically talking about that firmware version, but I did find several other reviews mentioning the issue, e.g. [this one](https://cryptonewsreview.com/review-coinkite-coldcard-bitcoin-and-litecoin-hardware-wallet/)

> Sadly, using the Coldcard – even at this stage – isn’t as easy as it could be. The keypad is extremely problematical: sometimes it suffers from debounce, registering two keypresses where there should be only one; sometimes it suffers from a lack of sensitivity, ignoring your presses entirely; other times it registered phantom presses, often scrolling to the bottom of the menu even while not being held.

**Comment by nvk** [comment]
-retracted- 

Regardless, should we delist Ledger to leaking user info? Or trezor with the numerous issues trough time?

The fee thing is preference. They keyboard can be improved but none of this reasons should prevent it from being listed.

**Comment by achow101** [comment]
Regarding the sensitivity issue with spurious keypresses, users can change the sensitivity of the keypad. I believe I had done this before which may explain the discrepancy between @crwatkins experience and mine.

I have also pulled down and built the development branch which has the next release of the firmware. I tested this and found that there are additional sensitivity settings which helps with the keypad issues and there is indeed a setting for changing the fee limit and it can be disabled completely.

In my testing and usage of the device, I never ran into the fee limit and actually hadn't known that it existed until @crwatkins pointed it out in the bitcoin.org review of the Coldcard. Granted, I had not been testing with very small amounts.

In general, I think this device is something that can be recommended to lesser experienced users. Its features and settings are fairly well documented, so those who experience issues such as spurious keypad presses should be able to find documentation describing how they can change settings to fix these problems. As for the fee limit issue, I don't think that will be something that people run into in practice very much, and the publicly available code does indicate that this can be disabled in the next firmware release.

**Comment by harding** [comment]
NACK this PR.  I suggest reconsideration be possible after three months (around 25 Februrary 2018).  Reason for NACK is that I don't believe the wallet passes the following of our listed criteria: "No indication that security issues have been concealed, ignored, or not addressed correctly in order to prevent new or similar issues from happening in the future."  Details below in reply to @nvk 

> That guy hates us, he also gave Opendime a bad review because we didn’t want to give him free products.

He may very well hate you, especially after you've accused him of graft, but the [review](https://github.com/bitcoin-dot-org/bitcoin.org/pull/2612#issuecomment-433630363) he gave the Coldcard demonstrated careful disinterested testing that revealed several significant issues.  This is par for course for @crwatkins, who has identified and responsibly disclosed bugs that would've likely resulted in loss or theft of bitcoins in multiple wallets over the past three years.  (And who has a respected history of performing vendor-neutral testing outside of Bitcoin as well.)

Although I don't think your documented issues are deal-breakers themselves, especially since there are (or soon will be) configuration options that can help address them, I do think that you've acted inappropriately according to our listing criteria by attempting to deflect from your issues by accusing a reviewer of malfeasance.  For the sake of people who will store thousands or millions of dollars in their wallets, we need to know that you take security issues---like phantom button presses---very seriously.  Hopefully in three months, there will be more positive reviews of Coldcard and less indication that your preferred response to issues is criticizing the reviewer.

FYI: after @crwatkins completed his review of OpenDimes, he mailed me two of them (at his own small expense for the stamp) so that I could check some of his analysis.  This is hardly the behavior of someone who just wanted free stuff---but it is the behavior of a careful reviewer who is trying to perform the best possible evaluation.

**Comment by achow101** [comment]
>> That guy hates us, he also gave Opendime a bad review because we didn’t want to give him free products.
>
> He may very well hate you, especially after you've accused him of graft

I think he was referring to the author of [the other review you linked](https://cryptonewsreview.com/review-coinkite-coldcard-bitcoin-and-litecoin-hardware-wallet/), not @crwatkins

**Comment by nvk** [comment]
-retracted- 

@crwatkins did not nothing but great work.

Go ahead just close the PR

**Comment by harding** [comment]
@nvk my abject apologies for inferring the wrong subject of your pronoun.

**Comment by ghalfacree** [comment]
> That guy (Gareth) hates us, he also gave Opendime a bad review because we didn’t want to give him free products.

That's a lie, @nvk, and you know it. I bought a set of Opendimes for review with my own money, and the first one I tried spontaneously unlocked itself. I contacted you on Twitter and you blamed mishandling on my part for damaging it, then when I showed you high-resolution imagery demonstrating that the heatshrink and unlocking resistor were both intact told me to contact support. I did, and they completely ignored my message. You then took to badmouthing me on Twitter - and, apparently, GitHub - when people asked you about issues with your products.

I also bought the Coldcard with my own money. I stand by both reviews.

At no time *ever* have I requested free products, and I challenge you to prove otherwise or edit your post to retract that libellous claim.

**Comment by nvk** [comment]
I wont do the laundry here, its not the place. I retract my comment.

**Comment by nvk** [comment]
I went over the messages with you, and as much as I disagree with you on a number of issues and on the Opendime matter I got you confused with another Blogger. We unfortunately get "grafted" alot by bloggers in this space and I thought it was you. I apologize for that.

**Comment by ghalfacree** [comment]
> I went over the messages with you, and as much as I disagree with you on a number of issues and on the Opendime matter I got you confused with another Blogger. We unfortunately get "grafted" alot by bloggers in this space and I thought it was you. I apologize for that.

Thank you, @nvk, I appreciate the apology - and with a name like mine, I can honestly say it's not often I get mistaken for someone else!

For the record: I don't hate you, your company, or your products. If you were to launch versions which address the issues I, and others in this thread and elsewhere, have encountered, I'd happily give them a glowing review.

**Comment by nvk** [comment]
I'm not a reader of your blog, I bumped into once when support brought to my attention over the Opendime issue and only again when your other review was linked here. I confused you with a recent disingenuous reviewer. You'd be surprised as to how many Bloggers and Youtubers contact us asking or demanding product for review. And how many write reviews without ever seeing the product.

Any ways, this thread is more than beaten to death.

**Comment by ghalfacree** [comment]
@nvk I would have thought you would have remembered claiming I was ["bitter" and my review "shitty"](https://twitter.com/nvk/status/980818961443864576) when I documented the same Opendime-spontaneous-unlock issue in Custom PC Magazine earlier this year, but I guess not.

Anyway, best of luck with addressing the issues raised in this thread and getting your device listed!

**Comment by nvk** [comment]
Thanks for reminding me of that tweet, I stand by it.

**Comment by nvk** [comment]
Neither you or I have grace as a virtue, i think it's a good time to stop here.

**Comment by ghalfacree** [comment]
> Thanks for reminding me of that tweet, I stand by it.

Including the part where you say, and I quote, "the writer was not able to show support it was not his fault," despite me having emailed [a high-resolution macro shot of the device showing absolutely zero physical damage](https://twitter.com/ghalfacree/status/951499162142629888) and having never received a single reply?

Well, okay then. Good to have that in the public record.

**Pull request closed without merging**